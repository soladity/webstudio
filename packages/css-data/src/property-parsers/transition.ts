import * as csstree from "css-tree";
import type {
  InvalidValue,
  LayersValue,
  TupleValue,
  TupleValueItem,
  Unit,
} from "@webstudio-is/css-engine";

const cssTreeTryParseValue = (input: string) => {
  try {
    const ast = csstree.parse(input, { context: "value" });
    return ast;
  } catch {
    return undefined;
  }
};

export const parseTransition = (
  transition: string
): LayersValue | InvalidValue => {
  let tokenStream = transition.trim();
  tokenStream = tokenStream.endsWith(";")
    ? tokenStream.slice(0, -1)
    : tokenStream;

  const cleanupKeywords = ["transition:"];

  for (const cleanupKeyword of cleanupKeywords) {
    tokenStream = tokenStream.startsWith(cleanupKeyword)
      ? tokenStream.slice(cleanupKeyword.length).trim()
      : tokenStream;
  }

  const cssAst = cssTreeTryParseValue(tokenStream);

  if (cssAst === undefined) {
    return {
      type: "invalid",
      value: transition,
    };
  }

  const parsed = csstree.lexer.matchProperty("transition", cssAst);
  if (parsed.error) {
    return {
      type: "invalid",
      value: transition,
    };
  }

  const layers: TupleValue[] = [];

  csstree.walk(cssAst, (node) => {
    if (node.type === "Value") {
      const children = node.children;
      let layer: csstree.CssNode[] = [];

      for (const child of children) {
        layer.push(child);

        if (child.type === "Operator" || children.last === child) {
          const transition: TupleValueItem[] = [];

          for (let index = 0; index < layer.length; index++) {
            const item = layer[index];
            if (item.type === "Identifier") {
              transition.push({
                type: "keyword",
                value: item.name,
              });
            }

            if (item.type === "Dimension") {
              transition.push({
                type: "unit",
                value: Number(item.value),
                unit: item.unit as Unit,
              });
            }
          }

          layers.push({
            type: "tuple",
            value: transition,
          });
          layer = [];
          continue;
        }
      }
    }
  });

  return layers.length > 0
    ? {
        type: "layers",
        value: layers,
      }
    : { type: "invalid", value: transition };
};
