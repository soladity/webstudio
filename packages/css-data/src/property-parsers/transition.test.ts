import { describe, expect, test } from "@jest/globals";

import { parseTransition } from "./transition";

describe("parseTransition", () => {
  test("parses value and returns invalid when used a invalid color", () => {
    expect(parseTransition(`10px 10px 5px foo`)).toMatchInlineSnapshot(`
      {
        "type": "invalid",
        "value": "10px 10px 5px foo",
      }
    `);
  });

  test("parse value and returns multiple valid layers", () => {
    expect(parseTransition("color .2s, text-shadow .2s")).toMatchSnapshot(`
      {
        "type": "layers",
        "value": [
          {
            "type": "tuple",
            "value": [
              {
                "type": "keyword",
                "value": "color"
              },
              {
                "type": "unit",
                "value": 0.2,
                "unit": "s"
              }
            ]
          },
          {
            "type": "tuple",
            "value": [
              {
                "type": "keyword",
                "value": "text-shadow"
              },
              {
                "type": "unit",
                "value": 0.2,
                "unit": "s"
              }
            ]
          }
        ]
      }`);
  });

  test("parse value and returns valid layer", () => {
    expect(parseTransition("opacity 200ms ease 0s")).toMatchSnapshot(`
      {
        "type": "layers",
        "value": [
          {
            "type": "tuple",
            "value": [
              {
                "type": "keyword",
                "value": "opacity"
              },
              {
                "type": "unit",
                "value": 200,
                "unit": "ms"
              },
              {
                "type": "keyword",
                "value": "ease"
              },
              {
                "type": "unit",
                "value": 0,
                "unit": "s"
              }
            ]
          }
        ]
      }`);
  });
});
