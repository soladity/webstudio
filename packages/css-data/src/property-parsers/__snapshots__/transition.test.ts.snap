// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTransition parse value and returns multiple valid layers: 
      {
        "type": "layers",
        "value": [
          {
            "type": "tuple",
            "value": [
              {
                "type": "keyword",
                "value": "color"
              },
              {
                "type": "unit",
                "value": 0.2,
                "unit": "s"
              }
            ]
          },
          {
            "type": "tuple",
            "value": [
              {
                "type": "keyword",
                "value": "text-shadow"
              },
              {
                "type": "unit",
                "value": 0.2,
                "unit": "s"
              }
            ]
          }
        ]
      } 1`] = `
{
  "type": "layers",
  "value": [
    {
      "type": "tuple",
      "value": [
        {
          "type": "keyword",
          "value": "color",
        },
        {
          "type": "unit",
          "unit": "s",
          "value": 0.2,
        },
      ],
    },
    {
      "type": "tuple",
      "value": [
        {
          "type": "keyword",
          "value": "text-shadow",
        },
        {
          "type": "unit",
          "unit": "s",
          "value": 0.2,
        },
      ],
    },
  ],
}
`;

exports[`parseTransition parse value and returns valid layer: 
      {
        "type": "layers",
        "value": [
          {
            "type": "tuple",
            "value": [
              {
                "type": "keyword",
                "value": "opacity"
              },
              {
                "type": "unit",
                "value": 200,
                "unit": "ms"
              },
              {
                "type": "keyword",
                "value": "ease"
              },
              {
                "type": "unit",
                "value": 0,
                "unit": "s"
              }
            ]
          }
        ]
      } 1`] = `
{
  "type": "layers",
  "value": [
    {
      "type": "tuple",
      "value": [
        {
          "type": "keyword",
          "value": "opacity",
        },
        {
          "type": "unit",
          "unit": "ms",
          "value": 200,
        },
        {
          "type": "keyword",
          "value": "ease",
        },
        {
          "type": "unit",
          "unit": "s",
          "value": 0,
        },
      ],
    },
  ],
}
`;
